### DISCUSSION TITLE
How can I solve this equation in couple with tensor mechanics module?

### DISCUSSION THREAD

**Discussion by K-Komal-98** [idaholab/moose#24361] [Q&A Modules: Solid mechanics]
![image](https://github.com/idaholab/moose/assets/113084320/1b1ee0ec-f7b6-48ba-bdc0-68279a50b6b9)

Here sigma_h is the hydrostatic stress. The terms two terms are straightforward. I can use TimeDerivative and Diffusion kernel. How can I solve for the last term where there is derivative of hydrostatic stress?

**Comment by GiudGiud** [comment]
Hello

Is the stress a single-valued field (=variable) here?
Or a tensor?

Guillaume

**Reply by K-Komal-98** [reply]
sigma_h is an average of normal stress components of a stress tensor i.e. sigma_11, sigma_22, sigma_33. I think it can be defined as an AuxVariable. Am I right?

**Reply by GiudGiud** [reply]
AuxVariable will likely be problematic because you will miss the derivatives of the stress tensor component with regards to each variable, to form the Jacobian later down the line

A material property is likely the way to go here

**Reply by K-Komal-98** [reply]
Thanks! 
Is there any example file where I can look into?

**Reply by GiudGiud** [reply]
Your best best is the tensor mechanics module. 
I'm not sure which model in particular though

**Reply by K-Komal-98** [reply]
Hi @GiudGiud !
In this discussion (#22910), the hydrostatic stress has been taken as an aux variable. You have suggested to use a material property. Are these two different approaches?

**Reply by GiudGiud** [reply]
I think we suggested material property there too

**Comment by avtarsinghh1991** [comment]
Hello there

Follow this discussion (https://github.com/idaholab/moose/discussions/22910) for incorporating the hydrostatic stress coupling.



Thanks,
Avtar

**Reply by K-Komal-98** [reply]
Thanks!

**Comment by hugary1995** [comment]
In addition to what I suggested in #22910, there is a cheaper way of getting the gradient using a local projection. See for example this code snippet: https://github.com/hugary1995/eel/blob/main/src/materials/chemistry/ChemicalPotential.C#L56-L78

MOOSE doesn't support `ADRealEigenMatrix` and `ADRealEigenVector` yet, so I defined those utils in my app.This is the third or fourth time I see this question pop up, I can create a PR or a gist to demonstrate this capability. @GiudGiud this might be a good side project for an intern. Ideally, this would be a material object called something like `MaterialPropertyGradient` which takes a scalar-valued material property, does the local projection, and interpolates its gradients onto the qps.

**Reply by GiudGiud** [reply]
Let s create an issue to keep track of this.
I agree this would make a good project. Once we have interns started on their main project we can consider it

**Reply by hugary1995** [reply]
Sounds good. Let me know if anyone starts working on this. I could also have someone here do it.

**Reply by GiudGiud** [reply]
My interns wont be up and running for another month. 
@permcody might have a resource

**Reply by rh201** [reply]
Hi, are there any updates on MaterialPropertyGradient, or we still follow suggestion in @hugary1995 #22910 ? Thanks.

**Reply by GiudGiud** [reply]
no update. 
Feel free to contribute such an object. There is some recent efforts into local projections but that was for mesh online modification purposes not material property gradients.
